rapthor/lib/cluster.py

```diff
@@ -3,118 +3,178 @@
-    elif antenna == 'LBA':
-        # Memory usage in GB/timeslot/dir of a typical LBA observation
-        mem_usage_gb = 0.05
-    gb_per_solint = mem_usage_gb * solint_fast_timestep * ndir
-
-    # Try to make at least as many time chunks as there are nodes, but ensure that
-    # solint_fast_timestep a divisor of samplesperchunk (otherwise we could get a lot
-    # of solutions with less than the target time)
-    n_nodes = cluster_parset['max_nodes']
-    samplesperchunk = np.ceil(numsamples / n_nodes)
-    if mem_gb / gb_per_solint < 1.0:
-        old_solint_fast_timestep = solint_fast_timestep
-        solint_fast_timestep *= mem_gb / gb_per_solint
-        solint_fast_timestep = max(1, int(round(solint_fast_timestep)))
+    def get_mem_per_solution(target_timestep, ndir):
+        # Estimate the memory usage per solution
+        #
+        # Note: the numbers below were determined empirically from typical (Dutch-only)
+        # datasets
+        if antenna == 'HBA':
+            # Memory usage in GB/timeslot/dir of a typical HBA observation
+            mem_usage_gb = 0.2
+        elif antenna == 'LBA':
+            # Memory usage in GB/timeslot/dir of a typical LBA observation
+            mem_usage_gb = 0.05
+
+        # Return total usage plus a 10% safety buffer
+        return mem_usage_gb * target_timestep * ndir * 1.1
+
+    # Determine whether we need to adjust the solution interval to fit the solve in
+    # memory
+    mem_gb = get_available_memory()
+    gb_per_sol = get_mem_per_solution(target_timestep, ndir)
+    if mem_gb / gb_per_sol < 1.0:
+        solint = target_timestep * mem_gb / gb_per_sol
+        solint = max(1, int(round(solint)))
+    else:
+        solint = target_timestep
+    if solint < target_timestep:
         log.warn('Not enough memory available for fast-phase solve. Reducing solution '
-                 'time interval from {0} to {1}'.format(old_solint_fast_timestep,
-                                                        solint_fast_timestep))
-    while samplesperchunk % solint_fast_timestep:
-        samplesperchunk -= 1
-    if samplesperchunk < solint_fast_timestep:
-        samplesperchunk = solint_fast_timestep
-    target_time_chunksize = timepersample * samplesperchunk
+                 'time interval from {0} to {1} time slots'.format(target_timestep, solint))
 
-    return target_time_chunksize, solint_fast_timestep
+    # Determine the size of chunks to split the calibration into (to allow
+    # parallelization over nodes).
+    #
+    # Try to make at least as many chunks (over all observations) as there are
+    # nodes and ensure that the solint is a divisor of samples_per_chunk
+    # (otherwise we could get a lot of solutions with less than the target time)
+    target_numchunks = np.ceil(cluster_parset['max_nodes'] / numobs)
+    samples_per_chunk = int(np.ceil(numsamples / target_numchunks))
+    while samples_per_chunk % solint:
+        samples_per_chunk -= 1
```

Comment:
I'm confused here. I think you what you want is that `samples_per_chunk` is lowered to a value that is divisible by `solint`, right. Why not write `samples_per_chunk -= samples_per_chunk % solint`, or `samples_per_chunk = solint * (samples_per_chunk // solint)`, instead of the while-loop?

Reply:
True -- that is a bit cleaner. I'll change it.